File: CORE/src/test/java/com/variant/core/test/TargetingTest.java
package com.variant.core.test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;

import java.util.Random;

import org.junit.Test;

import com.variant.core.VariantCoreSession;
import com.variant.core.VariantCoreStateRequest;
import com.variant.core.exception.VariantRuntimeException;
import com.variant.core.exception.VariantRuntimeUserErrorException;
import com.variant.core.hook.HookListener;
import com.variant.core.hook.TestTargetingHook;
import com.variant.core.impl.VariantCore;
import com.variant.core.schema.ParserResponse;
import com.variant.core.xdm.Schema;
import com.variant.core.xdm.State;
import com.variant.core.xdm.Test.Experience;
import com.variant.core.xdm.impl.MessageTemplate;

public class TargetingTest extends BaseTestCore {

	static final int TRIALS = 500000;
	static final float DELTA_AS_FRACTION = .05f;

	/**
	 * Use the null event flusher and null session store, 
	 * because the test will be generating lots sessions and events.
	 */
	static {
		// Do nothing to save sessions to speed things up. We won't need them.
		injectorConfigAsResourceName = "/com/variant/core/conf/injector-session-store-null.json";
	}

	VariantCore core = rebootApi();
	
	/**
	 * Basic targeting
	 * @throws Exception
	 */
	@Test
	public void basicTest() throws Exception {
				
		String config = 
				"{                                                             \n" +
			    "   'states':[                                                 \n" +
			    "     {  'name':'state1',                                      \n" +
			    "        'parameters':{                                        \n" +
			    "           'path':'/path/to/state1'                           \n" +
			    "        }                                                     \n" +
			    "     },                                                       \n" +
			    "     {  'name':'state2',                                      \n" +
			    "        'parameters':{                                        \n" +
			    "           'path':'/path/to/state2'                           \n" +
			    "        }                                                     \n" +
			    "     }                                                        \n" +
			    "  ],                                                          \n" +
				"  'tests':[                                                   \n" +
			    "     {                                                        \n" +
			    "        'name':'test1',                                       \n" +
			    "        'experiences':[                                       \n" +
			    "           {                                                  \n" +
			    "              'name':'A',                                     \n" +
			    "              'weight':1 ,                                    \n" +
			    "              'isControl':true                                \n" +
			    "           },                                                 \n" +
			    "           {                                                  \n" +
			    "              'name':'B',                                     \n" +
			    "              'weight':2                                      \n" +
			    "           },                                                 \n" +
			    "           {                                                  \n" +
			    "              'name':'C',                                     \n" +
			    "              'weight':97                                     \n" +
			    "           }                                                  \n" +
			    "        ],                                                    \n" +
			    "        'onStates':[                                          \n" +
			    "           {                                                  \n" +
			    "              'stateRef':'state1',                            \n" +
			    "              'variants':[                                    \n" +
			    "                 {                                            \n" +
			    "                    'experienceRef': 'B',                     \n" +
			    "                    'parameters':{                            \n" +
			    "                       'path':'/path/to/state1/test1.B'       \n" +
			    "                    }                                         \n" +
			    "                 },                                           \n" +
			    "                 {                                            \n" +
			    "                    'experienceRef': 'C',                     \n" +
			    "                    'parameters':{                            \n" +
			    "                       'path':'/path/to/state1/test1.C'       \n" +
			    "                    }                                         \n" +
			    "                 }                                            \n" +
			    "              ]                                               \n" +
			    "           }                                                  \n" +
			    "        ]                                                     \n" +
			    "     }                                                        \n" +
			    "  ]                                                           \n" +
			    "}                                                             \n";
		
		ParserResponse response = core.parseSchema(config);
		if (response.hasMessages()) printMessages(response);
		assertFalse(response.hasMessages());
		Schema schema = core.getSchema();		
		State state = schema.getState("state1");
		com.variant.core.xdm.Test test = schema.getTest("test1");

		//
		// No targeting listener - distribution according to weights.
		//
		int[] counts = {0, 0, 0};
		for (int i = 0; i < TRIALS; i++) {
			VariantCoreSession ssn = core.getSession("foo" + i, true).getBody();
			VariantCoreStateRequest req = ssn.targetForState(state);
			String expName = req.getLiveExperience(test).getName();
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
			req.commit();
		} 
		verifyCounts(counts, new float[] {1, 2, 97});
		
		//
		// Add one null listener - still distribution according to weights.
		//
		NullTargetingHookListener nullListener1 = new NullTargetingHookListener();
		nullListener1.postCount = 0;
		core.addHookListener(nullListener1);
		counts[0] = counts[1] = counts[2] = 0;
		for (int i = 0; i < TRIALS; i++) {
			VariantCoreSession ssn = core.getSession("foo" + i, true).getBody();
			VariantCoreStateRequest req = ssn.targetForState(state);
			String expName = req.getLiveExperience(test).getName();			
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
			req.commit();
		} 
		assertEquals(TRIALS, nullListener1.postCount);
		verifyCounts(counts, new float[] {1, 2, 97});

		//
		// Add two null listeners - still distribution according to weights.
		//
		NullTargetingHookListener nullListener2 = new NullTargetingHookListener();
		core.addHookListener(nullListener2);
		counts[0] = counts[1] = counts[2] = 0;
		nullListener2.postCount = nullListener1.postCount = 0;
		for (int i = 0; i < TRIALS; i++) {
			VariantCoreSession ssn = core.getSession("foo" + i, true).getBody();
			VariantCoreStateRequest req = ssn.targetForState(state);
			String expName = req.getLiveExperience(test).getName();			
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
			req.commit();
		} 
		assertEquals(TRIALS, nullListener1.postCount);
		assertEquals(TRIALS, nullListener2.postCount);
		verifyCounts(counts, new float[] {1, 2, 97});

		//
		// Add the A/B listener - changes distribution to 1/1/0.
		//
		ABTargetingHookListener ABListener = new ABTargetingHookListener();
		core.addHookListener(ABListener);
		counts[0] = counts[1] = counts[2] = 0;
		ABListener.postCount = nullListener2.postCount = nullListener1.postCount = 0;
		for (int i = 0; i < TRIALS; i++) {
			VariantCoreSession ssn = core.getSession("foo" + i, true).getBody();
			VariantCoreStateRequest req = ssn.targetForState(state);
			String expName = req.getLiveExperience(test).getName();			
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
			req.commit();
		} 
		assertEquals(TRIALS, nullListener1.postCount);
		assertEquals(TRIALS, nullListener2.postCount);
		assertEquals(TRIALS, ABListener.postCount);
		verifyCounts(counts, new float[] {1, 1, 0});

		//
		// Clear all listeners, then add the A/C listener - changes distribution to 1/0/1.
		//
		core.clearHookListeners();
		ACTargetingHookListener ACListener = new ACTargetingHookListener();
		core.addHookListener(ACListener);
		counts[0] = counts[1] = counts[2] = 0;		
		for (int i = 0; i < TRIALS; i++) {
			VariantCoreSession ssn = core.getSession("foo" + i, true).getBody();
			VariantCoreStateRequest req = ssn.targetForState(state);
			String expName = req.getLiveExperience(test).getName();	
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
			req.commit();
		} 
		assertEquals(TRIALS, ACListener.postCount);
		verifyCounts(counts, new float[] {1, 0, 1});

		//
		// Add null listener - should not change the distribution.
		//
		core.addHookListener(nullListener1);
		nullListener1.postCount = ACListener.postCount = 0;
		counts[0] = counts[1] = counts[2] = 0;		
		for (int i = 0; i < TRIALS; i++) {
			VariantCoreSession ssn = core.getSession("foo" + i, true).getBody();
			VariantCoreStateRequest req = ssn.targetForState(state);
			String expName = req.getLiveExperience(test).getName();	
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
			req.commit();
		} 
		assertEquals(TRIALS, nullListener1.postCount);
		assertEquals(TRIALS, ACListener.postCount);
		verifyCounts(counts, new float[] {1, 0, 1});

		//
		// Clear all listeners, then add the A/B/Null custom targeter + A/C targeter - changes distribution to 50/25/25.
		//
		core.clearHookListeners();
		ABNullTargetingHookListener ABNullListener = new ABNullTargetingHookListener();
		core.addHookListener(ABNullListener);
		core.addHookListener(ACListener);
		counts[0] = counts[1] = counts[2] = 0;
		ABNullListener.postCount = ACListener.postCount = 0;
		for (int i = 0; i < TRIALS; i++) {
			VariantCoreSession ssn = core.getSession("foo" + i, true).getBody();
			VariantCoreStateRequest req = ssn.targetForState(state);
			String expName = req.getLiveExperience(test).getName();	
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
			req.commit();
		} 
		assertEquals(TRIALS, ABNullListener.postCount);
		assertEquals(TRIALS, ACListener.postCount);
		verifyCounts(counts, new float[] {50, 25, 25});
	}
	
	/**
	 * 
	 * @param counts
	 * @param weights
	 */
	private static void verifyCounts(int[] counts, float[] weights) {
		float sumCounts = 0, sumWeights = 0;
		for (int i = 0; i < counts.length; i++) {
			sumCounts += counts[i];
			sumWeights += weights[i];
		}
		for (int i = 0; i < counts.length; i++) {
			//System.out.println("Delta: " + (weights[i]/sumWeights * DELTA_AS_FRACTION));
			assertEquals(weights[i]/sumWeights * DELTA_AS_FRACTION, counts[i]/sumCounts, weights[i]/sumWeights);
		}
	}

	/**
	 * Basic targeting
	 * @throws Exception
	 */
	@Test
	public void noWeightsExceptionTest() throws Exception {

		String config = 
				"{                                                             \n" +
			    "   'states':[                                                 \n" +
			    "     {  'name':'state1',                                      \n" +
			    "        'parameters':{                                        \n" +
			    "           'path':'/path/to/state1'                           \n" +
			    "        }                                                     \n" +
			    "     },                                                       \n" +
			    "     {  'name':'state2',                                      \n" +
			    "        'parameters':{                                        \n" +
			    "           'path':'/path/to/state2'                           \n" +
			    "        }                                                     \n" +
			    "     }                                                        \n" +
			    "  ],                                                          \n" +
				"  'tests':[                                                   \n" +
			    "     {                                                        \n" +
			    "        'name':'test1',                                       \n" +
			    "        'experiences':[                                       \n" +
			    "           {                                                  \n" +
			    "              'name':'A',                                     \n" +
			    "              'weight':1 ,                                    \n" +
			    "              'isControl':true                                \n" +
			    "           },                                                 \n" +
			    "           {                                                  \n" +
			    "              'name':'B'                                      \n" +
/*			    "              'weight':2                                      \n" + */
			    "           },                                                 \n" +
			    "           {                                                  \n" +
			    "              'name':'C',                                     \n" +
			    "              'weight':97                                     \n" +
			    "           }                                                  \n" +
			    "        ],                                                    \n" +
			    "        'onStates':[                                          \n" +
			    "           {                                                  \n" +
			    "              'stateRef':'state1',                            \n" +
			    "              'variants':[                                    \n" +
			    "                 {                                            \n" +
			    "                    'experienceRef': 'B',                     \n" +
			    "                    'parameters':{                            \n" +
			    "                       'path':'/path/to/state1/test1.B'       \n" +
			    "                    }                                         \n" +
			    "                 },                                           \n" +
			    "                 {                                            \n" +
			    "                    'experienceRef': 'C',                     \n" +
			    "                    'parameters':{                            \n" +
			    "                       'path':'/path/to/state1/test1.C'       \n" +
			    "                    }                                         \n" +
			    "                 }                                            \n" +
			    "              ]                                               \n" +
			    "           }                                                  \n" +
			    "        ]                                                     \n" +
			    "     }                                                        \n" +
			    "  ]                                                           \n" +
			    "}                                                             \n";
		
		core.clearHookListeners();
		ParserResponse response = core.parseSchema(config);
		if (response.hasMessages()) printMessages(response);
		assertFalse(response.hasMessages());
		Schema schema = core.getSchema();		
		final State state = schema.getState("state1");
		final VariantCoreSession ssn = core.getSession("foo", true).getBody();

		new ExceptionInterceptor<VariantRuntimeException>() {
			@Override public void toRun() { ssn.targetForState(state); }
			@Override public void onThrown(VariantRuntimeException e) { 
				assertEquals(e.getMessage(), new VariantRuntimeUserErrorException(MessageTemplate.RUN_WEIGHT_MISSING, "test1", "B").getMessage()); 
			}
			@Override public Class<VariantRuntimeException> getExceptionClass() { return VariantRuntimeException.class; }
		}.assertThrown();

	}
	
	/**
	 * Always return null.
	 */
	private static class NullTargetingHookListener implements HookListener<TestTargetingHook> {
		
		private int postCount = 0;
		
		@Override
		public Class<TestTargetingHook> getHookClass() {
			return TestTargetingHook.class;
		}
		
		@Override
		public void post(TestTargetingHook hook) {
			postCount++;
			// do nothing.
		}
	}

	/**
	 * returns A or B in equal probabilities.
	 */
	private static class ABTargetingHookListener implements HookListener<TestTargetingHook> {

		private int postCount = 0;
		private static Random rand = new Random(System.currentTimeMillis());

		@Override
		public Class<TestTargetingHook> getHookClass() {
			return TestTargetingHook.class;
		}
		
		@Override
		public void post(TestTargetingHook hook) {
			postCount++;
			com.variant.core.xdm.Test test = hook.getTest();
			Experience experience = rand.nextBoolean() ? test.getExperience("A") : test.getExperience("B");
			hook.setTargetedExperience(experience);
		}
	}

	/**
	 * returns A or C in equal probabilities.
	 */
	private static class ACTargetingHookListener implements HookListener<TestTargetingHook> {
		
		private int postCount = 0;
		private static Random rand = new Random(System.currentTimeMillis());
		
		@Override
		public Class<TestTargetingHook> getHookClass() {
			return TestTargetingHook.class;
		}
		
		@Override
		public void post(TestTargetingHook hook) {
			postCount++;
			com.variant.core.xdm.Test test = hook.getTest();
			if (test.getName().equals("test1") && hook.getTargetedExperience() == null) {
				Experience experience = rand.nextBoolean() ? test.getExperience("A") : test.getExperience("C");
				hook.setTargetedExperience(experience);
			}
		}
	}

	/**
	 * returns A 25% of the time, B 25% of the time and null 50% of the time..
	 */
	private static class ABNullTargetingHookListener implements HookListener<TestTargetingHook> {
		
		private int postCount = 0;
		private static Random rand = new Random(System.currentTimeMillis());
		
		@Override
		public Class<TestTargetingHook> getHookClass() {
			return TestTargetingHook.class;
		}
		
		@Override
		public void post(TestTargetingHook hook) {
			postCount++;
			com.variant.core.xdm.Test test = hook.getTest();
			if (test.getName().equals("test1") && hook.getTargetedExperience() == null) {
				Experience experience = rand.nextBoolean() ? (rand.nextBoolean() ? test.getExperience("A") : test.getExperience("B")) : null;
				hook.setTargetedExperience(experience);
			}
		}

	}
	
	
}

File: SERVER/test/com/variant/server/test/TargetingTest.scala
package com.variant.server.test

import com.variant.server.impl.SessionImpl
import com.variant.server.test.hooks.TestTargetingHook
import com.variant.server.test.hooks.TestTargetingHookNil
import com.variant.server.test.util.ParameterizedString
import com.variant.server.api.ServerException
import com.variant.server.boot.ServerErrorLocal
import com.variant.server.schema.SchemaDeployerString

class TargetingTest extends BaseSpecWithServer {

	val trials = 500000
	val deltaAsFraction = .05f

		

   val schemaJson = """
{
   'meta':{
      'name':'TargetingTest'
   },
   'states':[
      {
         'name':'state1',
         'parameters':{
            'path':'/path/to/state1'
         }
      },
      {
         'name':'state2',
         'parameters':{
            'path':'/path/to/state2'
         }
      } 
   ],
   'tests':[
      {
         'name':'test1',
         'hooks':[${hooks:}], // Defaults to null list
         'experiences':[ 
            {
               'name':'A', 
               'weight':1 , 
               'isControl':true 
            }, 
            {  
               'name':'B',
               'weight':2  
            },  
            {  
               'name':'C',
               'weight':97 
            }
         ],
         'onStates':[ 
            { 
               'stateRef':'state1',
               'variants':[ 
                  {
                     'experienceRef': 'B',
                     'parameters':{
                        'path':'/path/to/state1/test1.B'
                     }
                  },
                  { 
                     'experienceRef': 'C',
                     'parameters':{ 
                        'path':'/path/to/state1/test1.C'
                     }
                  }
               ] 
            }
         ]
      },
      {
         'name':'test2',
         'experiences':[ 
            {
               'name':'A', 
               'weight':1 , 
               'isControl':true 
            }, 
            {  
               'name':'B',
               'weight':2  
            },  
            {  
               'name':'C',
               'weight':97 
            }
         ],
         'onStates':[ 
            { 
               'stateRef':'state1',
               'variants':[ 
                  {
                     'experienceRef': 'B',
                     'parameters':{
                        'path':'/path/to/state1/test2.B'
                     }
                  },
                  { 
                     'experienceRef': 'C',
                     'parameters':{ 
                        'path':'/path/to/state1/test2.C'
                     }
                  }
               ] 
            }
         ]
      } 
   ]
}"""

	"Runtime" should {

      "target according to weights with no targeting hooks" in {

         val schemaSrc = ParameterizedString(schemaJson).expand()
         
         val schemaDeployer = SchemaDeployerString(schemaSrc)
         server.useSchemaDeployer(schemaDeployer)
         val response = schemaDeployer.parserResponse
         server.schema.isDefined mustBe true
         val schema = server.schema.get
         val state = schema.getState("state1")
         val test = schema.getTest("test1")
   		
         val counts = Array(0, 0, 0)
   		for (i <- 1 to trials) {
   			val ssn = SessionImpl.empty("sid")
   			ssn.targetForState(state)
   			val expName = ssn.coreSession.getStateRequest.getLiveExperience(test).getName()
   			expName match {
   			   case "A" => counts(0) += 1
      			case "B" => counts(1) += 1
      			case "C" => counts(2) += 1
   			}
   		} 
   		verifyCounts(counts, Array(1f, 2f, 97f))
      }
      
   	"target according to weights with null targeting hook" in {

   	   val schemaSrc = ParameterizedString(schemaJson).expand(         
               "hooks" -> 
               """ {
                     'name' :'nullTargetingHook',
                     'class':'com.variant.server.test.hooks.TestTargetingHookNil'
                   }
               """)
               
         val schemaDeployer = SchemaDeployerString(schemaSrc)
         server.useSchemaDeployer(schemaDeployer)
         val response = schemaDeployer.parserResponse

         server.schema.isDefined mustBe true
         val schema = server.schema.get
         val state = schema.getState("state1")
         val test = schema.getTest("test1")

   		val counts = Array(0, 0, 0)
   		for (i <- 1 to trials) {
   			val ssn = SessionImpl.empty("sid" + i)
   			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe null
   			val req = ssn.targetForState(state)
   			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe "test1"
   			val expName = req.getLiveExperience(test).getName()
   			expName match {
   			   case "A" => counts(0) += 1
      			case "B" => counts(1) += 1
      			case "C" => counts(2) += 1
   			}
   		} 
   		verifyCounts(counts, Array(1f, 2f, 97f))
      }
     
      "target according to weights with two null targeting hooks" in {

         val schemaSrc = ParameterizedString(schemaJson).expand(         
               "hooks" -> 
               """ {
                     'name' :'nullTargetingHook1',
                     'class':'com.variant.server.test.hooks.TestTargetingHookNil'
                   },
                   {
                     'name' :'nullTargetingHook2',
                     'class':'com.variant.server.test.hooks.TestTargetingHookNil'
                   }
               """)
               
         val schemaDeployer = SchemaDeployerString(schemaSrc)
         server.useSchemaDeployer(schemaDeployer)
         val response = schemaDeployer.parserResponse

         server.schema.isDefined mustBe true
         val schema = server.schema.get
         val state = schema.getState("state1")
         val test = schema.getTest("test1")

   		val counts = Array(0, 0, 0)
   		for (i <- 1 to trials) {
   			val ssn = SessionImpl.empty("sid" + i)
   			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe null
   			val req = ssn.targetForState(state)
   			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe "test1 test1"
   			val expName = req.getLiveExperience(test).getName()
   			expName match {
   			   case "A" => counts(0) += 1
      			case "B" => counts(1) += 1
      			case "C" => counts(2) += 1
   			}
   		} 
   		verifyCounts(counts, Array(1f, 2f, 97f))
      }


      "target at 1/1/0 with the null hook and the A/B hook" in {

         
         val schemaSrc = ParameterizedString(schemaJson).expand(         
               "hooks" -> 
               """ {
                     'name' :'nullHook',
                     'class':'com.variant.server.test.hooks.TestTargetingHookNil'
                   },
                   {
                     'name' :'A_B_Hook',
                     'class':'com.variant.server.test.hooks.TestTargetingHook',
                     'init': {'weights':[1,1,0]}
                   }
               """)
               
         val schemaDeployer = SchemaDeployerString(schemaSrc)
         server.useSchemaDeployer(schemaDeployer)
         val response = schemaDeployer.parserResponse

         server.schema.isDefined mustBe true
         val schema = server.schema.get
         val state = schema.getState("state1")
         val test = schema.getTest("test1")

   		val counts = Array(0, 0, 0)
   		for (i <- 1 to trials) {
   			val ssn = SessionImpl.empty("sid" + i)
   			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe null
   			ssn.getAttribute(TestTargetingHook.ATTR_KEY) mustBe null
   			val req = ssn.targetForState(state)
   			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe "test1"
   			ssn.getAttribute(TestTargetingHook.ATTR_KEY) mustBe "test1"
   			val expName = req.getLiveExperience(test).getName()
   			expName match {
   			   case "A" => counts(0) += 1
      			case "B" => counts(1) += 1
      			case "C" => counts(2) += 1
   			}
   		} 
   		verifyCounts(counts, Array(1f, 1f, 0f))
      }

		"still target at 1/1/1 with the null hook and the A/B/C hook" in {
         
         val schemaSrc = ParameterizedString(schemaJson).expand(         
               "hooks" -> 
               """ {
                     'name' :'nullHook',
                     'class':'com.variant.server.test.hooks.TestTargetingHookNil'
                   },
                   {
                     'name' :'A_B_CHook',
                     'class':'com.variant.server.test.hooks.TestTargetingHook',
                     'init': {'weights':[1,1,1]}
                   }
               """)
               
         val schemaDeployer = SchemaDeployerString(schemaSrc)
         server.useSchemaDeployer(schemaDeployer)
         val response = schemaDeployer.parserResponse

         server.schema.isDefined mustBe true
         val schema = server.schema.get
         val state = schema.getState("state1")
         val test = schema.getTest("test1")

   		val counts = Array(0, 0, 0)
   		for (i <- 1 to trials) {
   			val ssn = SessionImpl.empty("sid" + i)
   			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe null
   			ssn.getAttribute(TestTargetingHook.ATTR_KEY) mustBe null
   			val req = ssn.targetForState(state)
   			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe "test1"
   			ssn.getAttribute(TestTargetingHook.ATTR_KEY) mustBe "test1"
   			val expName = req.getLiveExperience(test).getName()
   			expName match {
   			   case "A" => counts(0) += 1
      			case "B" => counts(1) += 1
      			case "C" => counts(2) += 1
   			}
   		} 
   		verifyCounts(counts, Array(1f, 1f, 1f))
		}
		
	   "Throw exception if targeting hook sets bad experience" in {

         val schemaSrc = ParameterizedString(schemaJson).expand(         
               "hooks" -> 
               """ {
                     'name' :'nullHook',
                     'class':'com.variant.server.test.hooks.TestTargetingHookNil'
                   },
                   {
                     'name' :'A_B_CHook',
                     'class':'com.variant.server.test.hooks.TestTargetingHook',
                     'init': {'experience':'test2.A'}
                   }
               """)
               
         val schemaDeployer = SchemaDeployerString(schemaSrc)
         server.useSchemaDeployer(schemaDeployer)
         val response = schemaDeployer.parserResponse
         
         server.schema.isDefined mustBe true
         val schema = server.schema.get

         val state1 = schema.getState("state1")
		   
	
		   var ssn = SessionImpl.empty(newSid())
   	   
   	   val caughtEx = intercept[ServerException.User] {
             ssn.targetForState(state1)   // targeting hook returns an experience not from test1
         }
         caughtEx.getMessage mustBe (
                     new ServerException.User(
                           ServerErrorLocal.HOOK_TARGETING_BAD_EXPERIENCE, classOf[TestTargetingHook].getName, "test1", "test2.A"
                     ).getMessage)
 
	   }

	}
	
	/**
	 * 
	 * @param counts
	 * @param weights
	 */
	private def verifyCounts(counts: Array[Int], weights: Array[Float]) {
	   println(counts.mkString(","))
	   println(weights.mkString(","))
		var sumCounts = 0
		var sumWeights = 0f
		for (i <- 0 until counts.length) {
			sumCounts += counts(i)
			sumWeights += weights(i)
		}
		for (i <- 0 until counts.length) {
			//System.out.println("Delta: " + (weights[i]/sumWeights * DELTA_AS_FRACTION));
		   val countFraction = counts(i)/sumCounts.asInstanceOf[Float]
		   // +- will complain if weightsFraction is 0: if so, give it a small positive value instead
		   val weightsFraction = weights(i)/sumWeights.asInstanceOf[Float]

		   if (weightsFraction == 0)
      		countFraction mustEqual 0		      
		   else
      		countFraction mustEqual (weightsFraction +- weightsFraction * deltaAsFraction)
		}
	}
}
